sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/ReduceNumShufflePartitions.scala

```diff
@@ -180,25 +180,45 @@ case class ReduceNumShufflePartitions(conf: SQLConf) extends Rule[SparkPlan] {
 
 case class CoalescedShuffleReaderExec(
     child: QueryStageExec,
-    partitionStartIndices: Array[Int]) extends UnaryExecNode {
+    partitionStartIndices: Array[Int],
+    var isLocal: Boolean = false) extends UnaryExecNode {
```

Comment:
`ReduceNumShufflePartitions` and local shuffle reader are two different optimizations, and they are conflicting:
`ReduceNumShufflePartitions` adjusts the numPartitions by assuming the partitions are post-shuffle partitions. Their data size depends on the shuffle blocks they need to read. If we change the shuffle to local shuffle readers, then the partitions become pre-shuffle partitions, and their data size is different.

Reply:
`If we change the shuffle to local shuffle reader, then the partitions become pre-shuffle partitions, and their data size is different.`
@cloud-fan Here the local shuffle reader is still optimize the post-shuffle partitions. And I don't understand why the partitions become pre-shuffle partitions?

Reply:
Ok, Got it. In order to make code more clear, I will create `LocalShuffleReaderExec` later. Thanks.